Sming/sming.mk

```diff
@@ -0,0 +1,662 @@
+
+ESPTOOL_FLAGS   := -p $(COM_PORT) -b $(COM_SPEED_ESPTOOL)
+ESPTOOL_IMGFLAGS:= -ff $(SPI_SPEED)m -fm $(SPI_MODE) -fs $(SPI_SIZE_M)
+
+#==============================================================================
+#   source, build and include directories
+#------------------------------------------------------------------------------
+ifeq ($(ENABLE_GDB), 1)
+	MODULES         += $(THIRD_PARTY_DIR)/gdbstub
+endif
+
+SRC_DIR         := $(sort $(MODULES) $(patsubst %/,%,$(dir $(EXTRA_SRC))))
+BUILD_DIR       := $(addprefix $(BUILD_BASE)/,$(SRC_DIR))
+
+EXTRA_INCDIR    += $(SMING_HOME)/include $(SMING_HOME)/ $(SMING_HOME)/system/include
+EXTRA_INCDIR    += $(SMING_HOME)/Wiring $(SMING_HOME)/Libraries $(SMING_HOME)/SmingCore
+EXTRA_INCDIR    += $(SMING_HOME)/Services/SpifFS
+EXTRA_INCDIR    += $(THIRD_PARTY_DIR)/spiffs/src
+EXTRA_INCDIR    += $(THIRD_PARTY_DIR)/rboot $(THIRD_PARTY_DIR)/rboot/appcode
+
+#==============================================================================
+#   intermediate files
+#------------------------------------------------------------------------------
+#   name for the target project
+TARGET          = app
+
+APP_AR          := $(BUILD_BASE)/$(TARGET)_app.a
+TARGET_OUT_0    := $(BUILD_BASE)/$(TARGET)_0.out
+TARGET_OUT_1    := $(BUILD_BASE)/$(TARGET)_1.out
+
+#==============================================================================
+#   wifi settings
+#------------------------------------------------------------------------------
+#   use global wifi settings from eclipse environment variables, if possible
+WIFI_SSID       ?= ""
+WIFI_PWD        ?= ""
+
+ifneq ($(WIFI_SSID), "")
+	DEFINES += WIFI_SSID WIFI_PWD
+endif
+
+#==============================================================================
+#   sming debug output
+#------------------------------------------------------------------------------
+#   let debugf print print filenames and line numbers
+DEBUG_PRINT_FILENAME_AND_LINE ?= 0
+
+#   debug verbose level (DEBUG=3 INFO=2 WARNING=1 ERROR=0)
+DEBUG_VERBOSE_LEVEL ?= 2
+
+#==============================================================================
+#   clean-up and rebuild
+#------------------------------------------------------------------------------
+.PHONY: all clean rebuild
+
+all: dirs $(CUSTOM_TARGETS) $(FLASH_FILES) $(OTA_FILES)
+
+clean:
+	$(Q) rm -rf $(BUILD_BASE) $(FW_BASE) $(CUSTOM_TARGETS)
+
+rebuild: clean all
+
+#   disable build-in rules
+.SUFFIXES:
+
+#==============================================================================
+#   create required directories
+#------------------------------------------------------------------------------
+.phony: dirs
+
+$(BUILD_DIR) $(FW_BASE):
+	$(Q) mkdir -p $@
+
+dirs: $(BUILD_DIR) $(FW_BASE)
+
+#==============================================================================
+#   (re-)compile sming framework if required
+#------------------------------------------------------------------------------
+LIBSMING        := sming
```

Comment:
Well, I always prefer ':=' if possible. It has less confusing effects (more conventional logic, no "backward assignments") for occasional make users.